runtime/j9vm/jvm.c

```diff
@@ -1817,6 +1818,13 @@ jint JNICALL JNI_CreateJavaVM(JavaVM **pvm, void **penv, void *vm_args) {
 #else /* CALL_BUNDLED_FUNCTIONS_DIRECTLY */
 			zipFuncs = (J9ZipFunctionTable*) J9_GetInterface(IF_ZIPSUP, &j9portLibrary, j9binBuffer);
 #endif /* CALL_BUNDLED_FUNCTIONS_DIRECTLY */
+#ifdef WIN32
```

Comment:
Any reason not to do this for all platforms?

Reply:
Apparently this is a Windows platform specific problem, no similar problem reported in other platforms, nor did I reproduce the issue on Linux either. Java launcher native code is platform specific hence it might have difference encoding though I didn't go through those code.
 I think the change should be limited to the platform affected, i.e., Windows.

Reply:
Every platform has a platform specific encoding. There is no reason to limit the fix to Windows. Once the port library call is used, the code should be the same for all platforms.

Reply:
There is a cost to perform the conversion. It doesn't seem necessary unless such problem can be reproduced in another platform.

Reply:
I expect every platform will have the same issue, unless the platform encoding happens to be utf8. If we tried we could reproduce it everywhere.

Reply:
It wasn't reproduced in Linux, https://github.com/AdoptOpenJDK/openjdk-build/issues/1080 only called out Windows though.

Reply:
Further about the test of non-ascii character in the jar file path, `Ç` in a Window (`cp1252`) can reproduce the problem, i.e., `RI` can run with a jar in such file path but `OpenJ9` can't.

However this is not the case in Linux, tried a few machine and observed different locale settings:
Machine 1:
```
LANG=
LANGUAGE=
LC_CTYPE="POSIX"
LC_ALL=
```

Machine 2:
```
LANG=en_US
LC_CTYPE="C"
LC_ALL=C
```

Machine 3:
```
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_ALL=
```

Both `RI` and `OpenJ9` can't start with a jar file path containing non-ascii character in machine 1 & 2, and both `RI` and `OpenJ9` can run successfully at machine 3.

It appears Java launcher & JCL natives have separated code paths for Windows and others which is not surprising.
The goal here is to match `RI` behavior and fix the problems like #6722 in windows.

Reply:
> The goal here is to match RI behavior and fix the problems like #6722 in windows.

If the same changes fix other platforms, I don't see why it needs to be restricted to Windows just because Hotspot doesn't work.

Reply:
> If the same changes fix other platforms

There is no `same` changes for all platforms.

To fix #6722 windows problem, `j9str_convert(J9STR_CODE_MUTF8, J9STR_CODE_WINDEFAULTACP)` is needed in which `J9STR_CODE_WINDEFAULTACP` is Windows specific and mapped to win32 defination `CP_ACP`. (Note: this conversion will be added via an OMR PR)

There are 8 encoding supported by `j9str_convert` in one or two way conversion:
```
/* character set currently in effect */
#define J9STR_CODE_PLATFORM_RAW 1
/* modified UTF-8 */
#define J9STR_CODE_MUTF8 2
/* UTF-16 */
#define J9STR_CODE_WIDE 3
/* EBCDIC */
#define J9STR_CODE_EBCDIC 4
/* orthodox UTF-8 */
#define J9STR_CODE_UTF8 5
/* orthodox UTF-8 */
#define J9STR_CODE_LATIN1 6
/* Windows default ANSI code page */
#define J9STR_CODE_WINDEFAULTACP 7
/* Windows current thread ANSI code page */
#define J9STR_CODE_WINTHREADACP 8
```

`J9STR_CODE_PLATFORM_RAW` can't substitute `J9STR_CODE_WINDEFAULTACP` for Windows though the comment states _character set currently in effect_. 
It doesn't fix the problem in Linux platform when `RI` doesn't work.

When a Unicode character within a jar file path can't be recognized properly by current system encoding settings (i.e., showing question marks in a default system console), JVM is not expected to load jar files from such location.

If there is an encoding scenario that a Unicode is displayed properly but JVM can't load jar files from a path with that character, it is a bug to fix either for `RI` or `OpenJ9`. So far there is no such testcase against `RI` yet. `OpenJ9` can be fixed by the PR(s) proposed at Windows platform reported in #6722.
It doesn't seem necessary (and could be risky) to perform character conversion for other platforms without an actual testcase showing the problem, i.e., `RI` works but `OpenJ9` doesn't, or the Unicode character is displayed properly but not recognized by both JVMs.

Reply:
Updated the PR to use OMR API  `j9str_convert` for reviewing, keep `WIP` until adding a test.

Reply:
It appears not straightforward to add automated tests for this PR, can we start reviewing of this PR and  do such tests in a separated PR instead?